lessons 1-8

Kazachenko1998 opened the pull request:
test 1 (кроме 1 и 2)

Kazachenko1998 committed:
Сделано 4 задачи

Kazachenko1998 committed:
Сделано все кроме радиан

Kazachenko1998 committed:
кроме 1 и 2

gagarski commented on src/lesson1/task1/Simple.kt:
```diff
@@ -51,16 +51,19 @@ fun main(args: Array<String>) {
-fun seconds(hours: Int, minutes: Int, seconds: Int): Int = TODO()
-
+fun seconds(hours: Int, minutes: Int, seconds: Int): Int {
+return hours *3600+ minutes*60 +seconds
```
- Попробуйте применить автоматическое форматирование  в Intellij IDEA к данному участку кода (выделите этот участок и нажмите `Code -> Reformat Code` в меню). Посмотрите на результат и постарайтесь придерживаться такого стиля форматирования.
- [Здесь и далее] Функции, состоящие из одного `return` в Kotlin принято записывать в однострочной форме (см. пример `sqr` и `discriminant`)

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -69,7 +77,16 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
+    if ((rook==1)&&(bishop==1)) ss=3 else
+    if (bishop==1) ss=2 else
+         ss=1
+return ss
```
- Попробуйте применить автоматическое форматирование  в Intellij IDEA к данному участку кода (выделите этот участок и нажмите `Code -> Reformat Code` в меню). Посмотрите на результат и постарайтесь придерживаться такого стиля форматирования.
- Переменным лучше давать говорящие имена.

gagarski replied:
- Переменная `ss` здесь не очень нужна. Вы можете делать `return` прямо из веток ветвления.

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -69,7 +77,16 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
                           rookX: Int, rookY: Int,
-                          bishopX: Int, bishopY: Int): Int = TODO()
+                          bishopX: Int, bishopY: Int): Int {
+    var rook:Int=0; var bishop:Int=0; var ss:Int=0
```
`rook` и `bishop` у вас принимают значения 0 и 1, т. е. используются как булевские флаги. Используйте переменную типа`Boolean`

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -69,7 +77,16 @@ fun whichRookThreatens(kingX: Int, kingY: Int,
+    var rook:Int=0; var bishop:Int=0; var ss:Int=0
+    if ((rookX==kingX)or(rookY==kingY)) rook = 1
+    if (Math.abs(bishopX-kingX)==Math.abs(bishopY-kingY)) bishop = 1
+    if ((rook==0)&&(bishop==0)) ss=0 else
```
`else` в цепочке ветвлений принято переносить на новую строку вместе со следующим `if`:

``` kotlin
if (someCondition) someAction()
else if (someOtherCondition) someOtherAction()
else anotherAction()
```

или так:

``` kotlin
if (someCondition) 
    someAction()
else if (someOtherCondition) 
    someOtherAction()
else 
    anotherAction()
```

В таком случае среда разработки будет сама правильно форматировать код и вам не придётся двигать "лесенку" влево руками.

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -79,14 +96,30 @@ fun rookOrBishopThreatens(kingX: Int, kingY: Int,
+fun triangleKind(a: Double, b: Double, c: Double): Int {
+    var ss: Int = 0
+    if ((a + b > c) && (a + c > b) && (c + b > a)) {
+        if ((a * a + b * b < c * c) or (a * a + c * c < b * b) or (b * b + c * c < a * a)) ss = 2 else
```
См. выше про  `else if`

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -79,14 +96,30 @@ fun rookOrBishopThreatens(kingX: Int, kingY: Int,
  */
-fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
+fun triangleKind(a: Double, b: Double, c: Double): Int {
+    var ss: Int = 0
```
- Переменным лучше давать говорящие имена.

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -79,14 +96,30 @@ fun rookOrBishopThreatens(kingX: Int, kingY: Int,
\ No newline at end of file
+fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int  {
+    var s : Int = 0
+     if (c<=a && d>=b) s = (b - a)  else
```
`else if`...

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -79,14 +96,30 @@ fun rookOrBishopThreatens(kingX: Int, kingY: Int,
+          if (c<=a && d>=a && d<=b) s = (d-a)  else
+              if (c>=a && c<=b && d>=b) s = (b-c) else
+                  if (c>=a && c<=b && d<=b) s = (d-c) else s = -1
+return s
```
- Попробуйте применить автоматическое форматирование  в Intellij IDEA к данному участку кода (выделите этот участок и нажмите `Code -> Reformat Code` в меню). Посмотрите на результат и постарайтесь придерживаться такого стиля форматирования.

gagarski replied:
- См. комментарий про `else if`

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,8 +56,16 @@ fun timeForHalfWay(t1: Double, v1: Double,
+                       rookX2: Int, rookY2: Int): Int {
+    var rook:Int=0; var bishop:Int=0; var ss:Int=0
+    if ((rookX1==kingX)or(rookY1==kingY)) rook = 1
+    if ((rookX2==kingX)or(rookY2==kingY)) bishop = 1
```
В данной задаче нет никакого `bishop`'а.
Также см. комментарии из следующей задачи.

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -79,14 +96,30 @@ fun rookOrBishopThreatens(kingX: Int, kingY: Int,
  *
  * Даны четыре точки на одной прямой: A, B, C и D.
- * Координаты точек a, b, c, d соответственно, b >= a, d >= c.
+ * Координаты точек a, b, c, d соответственно, a <= b, c <= d.
```
- В коммитах лучше избегать ненужных изменений, так как они усложняют объединение наборов изменений от разных разработчиков.

ElenaLaufeyson committed:
Весь решенный 1 тест.

ElenaLaufeyson committed:
Весь решенный 1 тест.

Kazachenko1998 committed:
весь тест сделан(прошлые не менял)

Kazachenko1998 committed:
весь тест сделан

Kazachenko1998 committed:
весь тест сделан

Kazachenko1998 committed:
Сложные оставил на потом, что делать с sin and cos  не понял

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +58,72 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+    if (x != 0) return nod(y % x, x) else return y;
+}
+
+fun nok(x: Int, y: Int): Int = (x / nod(x, y)) * y
```
Вызов `lcm` состоит только из вызова `nok`. Нужен ли `nok`?

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +58,72 @@ fun digitCountInNumber(n: Int, m: Int): Int =
-fun lcm(m: Int, n: Int): Int = TODO()
+
+
+fun nod(x: Int, y: Int): Int {
```
По-английски НОД называется greatest common divisor (gcd). Старайтесь избегать использваония транслита в названиях функций/переменных.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +132,15 @@ fun maxDivisor(n: Int): Int = TODO()
  */
-fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+fun isCoPrime(m: Int, n: Int): Boolean {
+    val nn = Math.max(m, n)
```
Проверить числа на взаимную простоту можно в одну строчку с учётом того, что вы сделали ранее в этом уроке.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +132,15 @@ fun maxDivisor(n: Int): Int = TODO()
+    var i: Int = 0
+    var failed: Boolean = false
+    for (i in 2..nn) {
+        if ((n % i == 0) && (m % i == 0)) failed = true
```
[Будет неактуально, когда исправите замечание выше] Когда вы присвоили переменной `failed` значение `true`, можно завершить выполнение функции (сделав `return` прямо из тела цикла)

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +132,15 @@ fun maxDivisor(n: Int): Int = TODO()
+    for (i in 2..nn) {
+        if ((n % i == 0) && (m % i == 0)) failed = true
+    }
+    if (failed == true) return false else return true
```
[Будет неактуально, когда исправите замечания выше] Оператор `==` тоже возвращает булевское значение. 
Это ветвление **чрезвычайно** многословно для того, что оно делает.
В вашем случае его можно записать как
`return failed != true`
или ещё проще:
`return failed == false`
или совсем просто
`return !failed`

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -105,7 +149,15 @@ fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+        i++
+        if ((i * i >= m) && (i * i <= n)) failed = true
+    }
+    if (failed == true) return true else return false
```
То же самое, что и выше про ветвления и булеву алгебру.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -105,7 +149,15 @@ fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+    var failed: Boolean = false
+    while (i * i <= n) {
+        i++
+        if ((i * i >= m) && (i * i <= n)) failed = true
```
Так же, как и в предыдущей задаче, можно завершить выполнение, как только будет найдено подходящее значение. Тест падает из-за того, что вы перебираете все числа от 0 до `Integer.MAX_VALUE`, хотя можно завршить цикл на первой же итерации.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +166,8 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
  * Нужную точность считать достигнутой, если очередной член ряда меньше eps по модулю
  */
-fun sin(x: Double, eps: Double): Double = TODO()
+fun sin(x: Double, eps: Double): Double = Math.sin(x)
```
Вызов стандартной функции - не то, что предполагалось в этом задании. 
Нужно рассчитать синус по формуле, указанной в тексте путём постепенного суммирования членов ряда (`x`, `-x^3 / 3!` и т. д.). Когда останавливать суммирование так же сказано в тексте задания.

За математическим обоснованием формулы обращайтесь к лекциям по высшей математике или к гуглу по запросу "ряд Тейлора".

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +176,23 @@ fun sin(x: Double, eps: Double): Double = TODO()
  * Нужную точность считать достигнутой, если очередной член ряда меньше eps по модулю
  */
-fun cos(x: Double, eps: Double): Double = TODO()
+fun cos(x: Double, eps: Double): Double = Math.cos(x)
```
См. комментарий про синус.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +176,23 @@ fun sin(x: Double, eps: Double): Double = TODO()
-fun revert(n: Int): Int = TODO()
+fun revert(n: Int): Int {
+    var answer: Int = 0
+    var nn: Int = n
```
- Переменным лучше давать говорящие имена.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,15 +201,32 @@ fun revert(n: Int): Int = TODO()
-fun isPalindrome(n: Int): Boolean = TODO()
+fun isPalindrome(n: Int): Boolean {
+    var answer: Int = 0
+    var nn: Int = n
```
- Переменным лучше давать говорящие имена.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,15 +201,32 @@ fun revert(n: Int): Int = TODO()
+        answer = answer * 10 + nn % 10
+        nn = nn / 10
+    }
+    if (answer == n) return true else return false
```
См. замечания выше про ветвления и булевские функции.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,15 +201,32 @@ fun revert(n: Int): Int = TODO()
+        if (nn % 10 != memory) failed = false
+        nn = nn / 10
+    }
+    if (failed == false) return true else return false
```
См. замечания выше про ветвления и булевские функции.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,15 +201,32 @@ fun revert(n: Int): Int = TODO()
+    var nn: Int = n
+    var failed: Boolean = true
+    while (nn > 0) {
+        if (nn % 10 != memory) failed = false
```
См. замечания выше про досрочное завершение цикла.

Kazachenko1998 committed:
Lesson3.task1== All right

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -150,13 +152,9 @@ fun isCoPrime(m: Int, n: Int): Boolean {
-    if (failed == true) return true else return false
+    val k: Int = sqrt(n.toDouble()).toInt()
+
+    if ((k * k <= n) && (k * k >= m)) return true else return false
```
`return (k * k <= n) && (k * k >= m)`

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -245,14 +267,21 @@ fun quantity(n: Int): Int {
+fun squareSequenceDigit(n: Int): Int {
+    var i = 0
+    var number=0
+    var resalt=0
```
`result`

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -261,4 +290,17 @@ i++
+fun fibSequenceDigit(n: Int): Int {
+    var i = 0
+    var number=0
+    var resalt=0
```
`result`

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +164,20 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
  * Нужную точность считать достигнутой, если очередной член ряда меньше eps по модулю
  */
-fun sin(x: Double, eps: Double): Double = TODO()
+fun sin(x: Double, eps: Double): Double {
```
Попробуйте протестировать вашу функцию на больших `x`: 10, 100, 1000. Правильный ли результат она выдаёт? Попробуйте подумать, почему и какое свойство синуса можно использовать чтобы это исправить.

gagarski replied:
Исправлено

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +186,35 @@ fun sin(x: Double, eps: Double): Double = TODO()
  * Нужную точность считать достигнутой, если очередной член ряда меньше eps по модулю
  */
-fun cos(x: Double, eps: Double): Double = TODO()
+fun cos(x: Double, eps: Double): Double {
```
См. комментарий про синус.

gagarski replied:
Исправлено

gagarski commented:
Постарайтесь исправить текущие замечания и ошибки прежде чем приступать к следующим урокам.

gagarski commented on src/lesson2/task2/Logical.kt:
```diff
@@ -17,15 +18,21 @@ fun pointInsideCircle(x: Double, y: Double, x0: Double, y0: Double, r: Double) =
  */
-fun isNumberHappy(number: Int): Boolean = TODO()
+fun isNumberHappy(number: Int): Boolean {
+    if ((number % 10 + number / 10 % 10) == (number / 100 % 10 + number / 1000)) return true else return false
```
Здесь и далее вы используете довольно многословный способ вовзращения булевского значения.

``` kotlin
if (x) return true else return false
```

эквивалентно

``` kotlin
return x
```

gagarski commented on src/lesson2/task2/Logical.kt:
```diff
@@ -46,4 +55,6 @@ fun circleInside(x1: Double, y1: Double, r1: Double,
  */
-fun brickPasses(a: Int, b: Int, c: Int, r: Int, s: Int): Boolean = TODO()
+fun brickPasses(a: Int, b: Int, c: Int, r: Int, s: Int): Boolean {
+    if (((a <= r) && (b <= s)) || ((a <= s) && (b <= r)) || ((a <= r) && (c <= s)) || ((a <= s) && (a <= r)) || ((c <= r) && (b <= s)) || ((c <= s) && (b <= r))) return true else return false
```
Условие можно сильно упростить если заранее найти минимальные стороны кирпича.

Kazachenko1998 committed:
Lesson4.task1= все кроме последних 2 потому что очень много кода, хотя само задание не сложное

Kazachenko1998 committed:
Lesson4.task1= все кроме последних 2 потому что очень много кода, хотя само задание не сложное
не большие исправления в неверных задачах

Kazachenko1998 committed:
Lesson4.task1= все кроме последних 2 потому что очень много кода, хотя само задание не сложное
не большие исправления в неверных задачах

Kazachenko1998 committed:
Lesson5.task1 проверка на првильность

ElenaLaufeyson committed:
Исправила ошибки в первом уроке (синтаксис и неправильно решенную функцию angleInRadian), а также сделала 3 задания из второго урока.

ElenaLaufeyson committed:
Исправила ошибки из второго урока (сделаны только первые 3 задания).

ElenaLaufeyson committed:
Исправляю ошибки из второго урока...

Kazachenko1998 committed:
Lesson5.task1 кроме 2 последних

Kazachenko1998 committed:
Lesson5.task1 кроме 2 последних (испрвленны ошибки)

Kazachenko1998 committed:
Lesson5.task1 кроме 2 последних (испрвленны ошибки: телефон)

Kazachenko1998 committed:
Lesson5.task1 кроме 2 последних (испрвленны ошибки: телефон+ прыжок в высоту)

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -157,7 +257,31 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
+fun squareSequenceDigit(n: Int): Int {
+    var i = 0
+    var number=0
+    var resalt=0
```
`result`

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -1,7 +1,9 @@
 @file:Suppress("UNUSED_PARAMETER")
+
```
- В коммитах лучше избегать ненужных изменений, так как они усложняют объединение наборов изменений от разных разработчиков.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -193,7 +295,11 @@ fun convertToString(n: Int, base: Int): String = TODO()
-fun decimal(digits: List<Int>, base: Int): Int = TODO()
+fun decimal(digits: List<Int>, base: Int): Int {
+    var result: Int = 0
+    for (i in (digits.size - 1) downTo 0) result = result + (digits[i] * Math.pow(base.toDouble(), ((digits.size - 1) - i).toDouble())).toInt()
```
Старайтесь избегать таких длинных строк. IDEA рисует вертикальную линию на 120 символах. Старайтесь придерживаться хотя бы этого ограничения и переносить строки.

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +146,21 @@ fun dateDigitToStr(digital: String): String = TODO()
+fun flattenPhoneNumber(phone: String): String {
+    val result = phone.filter { (it in '0'..'9') || (it == '+') }
+
+    for (i in 0..phone.length - 1) if ((phone[i] != '(') && (phone[i] != ')') && (phone[i] != '-') && (phone[i] != ' ') && (phone[i] != '+') && (phone[i]!in '0'..'9')) return ""
```
См. комментарий про длинные строки.

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +172,16 @@ fun flattenPhoneNumber(phone: String): String = TODO()
  */
-fun bestLongJump(jumps: String): Int = TODO()
+fun bestLongJump(jumps: String): Int {
+    for (i in 0..jumps.length - 1) if ((jumps == "") || ((jumps[i] != '%') && (jumps[i] != '-') && (jumps[i] != ' ') && (jumps[i]!in '0'..'9'))) return -1
```
Длинные строки.

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +193,32 @@ fun bestLongJump(jumps: String): Int = TODO()
  * При нарушении формата входной строки вернуть -1.
  */
-fun bestHighJump(jumps: String): Int = TODO()
+fun chekPlus(str: String): Boolean {
```
`checkPlus`

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +229,26 @@ fun bestHighJump(jumps: String): Int = TODO()
  */
-fun plusMinus(expression: String): Int = TODO()
+fun plusMinus(expression: String): Int {
+    for (i in 0..expression.length - 1) if ((expression == "") || ((expression[i] != '+') && (expression[i] != '-') && (expression[i] != ' ') && (expression[i]!in '0'..'9'))) return throw IllegalArgumentException("Description")
```
Длинные строки.

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +229,26 @@ fun bestHighJump(jumps: String): Int = TODO()
+            else result = result + parts[i].toInt()
+        }
+    } catch (e: IllegalArgumentException) {
+        throw IllegalArgumentException("Description")
```
Вы ловите исключение `IllegalArgumentException` и тут же кидаете такое же, но с другим (бессмысленным) сообщением. Можно убрать этот `try`/`catch`.

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -69,7 +100,39 @@ fun dateStrToDigit(str: String): String = TODO()
+
+            if (parts[0].toInt() in 0..9) result = parts[0].toInt().toString() + " "
+            else result = parts[0].toInt().toString() + " "
+            result = when {
```
Можно использовать оператор `+=`. Также можно использовать конструкцию `when` с аргументом (для проверки на равенство).

``` kotlin
result += when (parts[1]) {
    "01" -> "января"
   // и т. д.
```

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +60,38 @@ fun main(args: Array<String>) {
+
+            if (parts[0].toInt() in 0..9) result = '0' + parts[0].toInt().toString() + "."
+            else result = parts[0].toInt().toString() + "."
+            result = when {
```
См. про `+=` и `when(...)` ниже.

ElenaLaufeyson committed:
Lesson2_task1
Исправлены ошибки в функциях timeForHalfWay и whichRookThreatens и решена функция rookOrBishopThreatens

ElenaLaufeyson committed:
Lesson2_task1
Сделана функция triangleKind.

ElenaLaufeyson committed:
Lesson2_task2
Сделана функция isNumberHappy.

ElenaLaufeyson committed:
Lesson2_task1
Доделан полностью.

ElenaLaufeyson committed:
Lesson2_task2
Доделала полностью.

ElenaLaufeyson committed:
Lesson3_task1
Решила, но еще не полностью задания 3 урока.

ElenaLaufeyson committed:
Lesson3_task1
Исправлена функция fun digitNumber.

ElenaLaufeyson committed:
Lesson3_task1
Исправлена тривиальная задача.

ElenaLaufeyson committed:
Lesson3_task1
Решила одну сложную задачу.

ElenaLaufeyson committed:
Lesson3_task1
Исправлена проблема с функцией sin и добавлена тестовая функция.

Kazachenko1998 committed:
Lesson3.task1 устранил замечания

Kazachenko1998 committed:
Lesson4.task1 устранил замечания

Kazachenko1998 committed:
Lesson5.task1 устранил замечания

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +193,48 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+    var s = ""
+    list = factorize(n)
+    for (i in 0..list.size - 1) {
+        if (i != list.size - 1) s = s + (list[i]).toString() + "*"
```
Есть метод `joinToString`

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -1,6 +1,8 @@
-@file:Suppress("UNUSED_PARAMETER")
+@file:Suppress("UNUSED_PARAMETER", "UNREACHABLE_CODE", "VARIABLE_WITH_REDUNDANT_INITIALIZER")
```
Старайтесь исправлять код с warning'ами, а не отключать их. Если вам действительно нужно их отключить (а это не такой случай), то старайтесь делать это более локально (например, для функции)

gagarski commented:
Остались ошибки при вычислении синуса/косинуса (см. комментарий) и замечание по `isCoPrime`.

gagarski commented:
recheck all

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +238,33 @@ fun bestHighJump(jumps: String): Int = TODO()
+                && (expression[i] != '-')
+                && (expression[i] != ' ')
+                && (expression[i] !in '0'..'9')))
+            return throw IllegalArgumentException("Description")
```
Чтобы бросить исключение нужно написать `throw IllegalArgumentException("Description")`

gagarski commented:
> как можно бросить исключение если мы его не ловили и не известно есть ли оно вообще?

Не понял вашего вопроса. Если комментарий выше не отвечает на него, то переформулируйте его:)

gagarski commented:
recheck all

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +158,20 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
+    var i = 0
+    var sinus: Double = x
+    var number: Double = x
+    while (Math.abs(number) * 1000 > eps) {
```
Вы таким способом неявно повышаете заданную точность. Вспомните свойства синуса.

gagarski replied:
Вызовите `sin(Math.PI / 2, 0.01)` и `sin(2000 * Math.PI + Math.PI / 2, 0.01)`. Правильные ли результаты возвращает функция?

gagarski commented:
> Остались ошибки при вычислении синуса/косинуса (см. комментарий) и замечание по isCoPrime.

Также остались замечания по 1-2 урокам

gagarski reviewed (changes requested):
`sin`, `cos`, `isCoPrime`, замечания по 1-2 урокам.

Kazachenko1998 committed:
Lesson6.task1 чуть-чуть для проверки

Kazachenko1998 committed:
Lesson1.task1 устранены замечания

gagarski commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -55,36 +55,61 @@ data class Circle(val center: Point, val radius: Double) {
      */
-    fun contains(p: Point): Boolean = TODO()
+    fun contains(p: Point): Boolean =
+            if ((sqr(center.x - p.x) + sqr(center.y - p.y)) <= sqr(radius)) true
```
Снова `if (...) true else false`.

gagarski commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -55,36 +55,61 @@ data class Circle(val center: Point, val radius: Double) {
+     * Дано множество точек. Вернуть отрезок, соединяющий две наиболее удалённые из них.
+     * Если в множестве менее двух точек, бросить IllegalArgumentException
+     */
+    fun distance(first : Point, second : Point): Double = Math.sqrt(sqr(first.x - second.x) + sqr(first.y - second.y))
```
Расстояние между точками уже реализовано в классе `Point`.

gagarski commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -55,36 +55,61 @@ data class Circle(val center: Point, val radius: Double) {
+    fun diameter(vararg points: Point): Segment {
+        var max = -1.0
+        var ii = 0
+        var jj = 0
```
Можно хранить не индексы, а сами точки. Операция присваивания не будет более медленной, поскольку при присваивании будут присваиваться ссылки на лочки.

gagarski commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -55,36 +55,61 @@ data class Circle(val center: Point, val radius: Double) {
+        if (points.size < 2) {
+            throw IllegalArgumentException()
+        }
+        for (i in 0..points.size-1)
```
Когда внутри `for` только одно выражение, можно записывать его без фигурных скобок. Но если это выражение неоднострочное (как другой `for` или `if`), то запись без скобок немного затрудняет понимание участка кода.

Kazachenko1998 committed:
Lesson2.task1 устранены замечания

Kazachenko1998 committed:
Lesson2.task2 устранены замечания

Kazachenko1998 commented:
есть еще какие-то замечания по 1-2 урокам?

gagarski commented:
> есть еще какие-то замечания по 1-3 урокам?
- Обратите внимание, что не проходят тесты для `ageDescription`  и `brickPasses`
- Когда будете чинить `brickPasses`, попробуйте таки найти две минимальные стороны кирпича заранее и посмотреть, насколько условие станет проще.
- Постарайтесь осознать, что когда вы пишете `if (x) true else false` или `if (x) return true else return false` это полностью эквивалентно `x` или `return x` соответственно (попробуйте убедиться в этом в уме или на бумаге для `x == true` и `x == false`). Вы везде пишете код вроде

``` kotlin
 fun circleInside(x1: Double, y1: Double, r1: Double,
                  x2: Double, y2: Double, r2: Double): Boolean = TODO()
                  x2: Double, y2: Double, r2: Double): Boolean =
         if ((Math.sqrt(sqr(x2 - x1) + sqr(y2 - y1)) + r1) <= r2) true else false
```

когда можно написать

``` kotlin
 fun circleInside(x1: Double, y1: Double, r1: Double,
                  x2: Double, y2: Double, r2: Double): Boolean = TODO()
                  x2: Double, y2: Double, r2: Double): Boolean =
         Math.sqrt(sqr(x2 - x1) + sqr(y2 - y1)) + r1) <= r2
```

(результат операции `<=` **уже** булевское значение). Исправлять это везде сейчас не нужно, но в будущем постарайтесь вспомниать об этом, если будете писать такие конструкции.
- Синус и косинус всё ещё сломаны для больших чисел.

Kazachenko1998 committed:
кажется все...

Kazachenko1998 committed:
мелкие исправления

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +237,33 @@ fun bestHighJump(jumps: String): Int = TODO()
+                else result += parts[i].toInt()
+        }
+    } catch (e: NumberFormatException) {
+        throw NumberFormatException("IllegalArgumentException")
```
Опечатка: видимо вы имели в виду `throw IllegalArgumentException("IllegalArgumentException")`

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -129,20 +274,51 @@ fun plusMinus(expression: String): Int = TODO()
- * Все цены должны быть положительными
- */
-fun mostExpensive(description: String): String = TODO()
+@Suppress("VARIABLE_WITH_REDUNDANT_INITIALIZER")
```
Лучше чем для всего файла, но всё же попробуйте понять, почему этот warning появляется.
Скорее всего он возникает из-за того, что вы объявляете переменную `string`, инициализируя её пустым списком: 

``` kotlin
var string = listOf<String>()
```

а затем, не используя это значение, меняете значение переменной:

``` kotlin
string = description.split("; ")
```

Следуя сообщению из warning'а, можно убрать инициализирующее значение, но тогда компилятор не сможет вывести тип переменной, поэтому нужно указать его вручную:

``` kotlin
var string: List<String>
```

После этого и warning'а не будет и область видимости переменной будет такая, как вам нужна.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +280,15 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
+    list = listOf()
+    var string1 = str
+    for (i in 0..string1.length - 1)
+        if (string1[i] in '0'..'9') list += ((string1[i]).toInt() - 48)
```
Лучше написать не 48, а `'0'.toInt()`, так будет понятно, что это за число.

gagarski commented:
> [Normal] lesson4.task1/center
> org.opentest4j.AssertionFailedError : You should mutate an input list, not create a copy
> [Normal] lesson4.task1/center
> org.opentest4j.AssertionFailedError : You should mutate an input list, not create a copy

Это сообщение говорит о том, что список поданный на вход функции должен быть изменён после её вызова

gagarski commented:
Есть ошибки (`plusMinus` и `center`, см. тесты)
По 1-5 уроку замечаний по коду, которые надо исправить нет. 
По 6 уроку есть замечание про `distance` и про индексы в `diameter`

ElenaLaufeyson committed:
Lesson3_task1
Доделала весь третий урок.

ElenaLaufeyson committed:
Lesson4_task1
Начала 4 урок.

Kazachenko1998 committed:
lesson6.task1 кроме очень сложных

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +129,7 @@ fun maxDivisor(n: Int): Int = TODO()
  * Например, 25 и 49 взаимно простые, а 6 и 8 -- нет.
  */
-fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+fun isCoPrime(m: Int, n: Int): Boolean = (nod(m, n) > Math.min(m, n)) || (nod(m, n) == 1)
```
Зачем первое условие в ИЛИ (`nod(m, n) > Math.min(m, n))`). Разве НОД двух чисел может быть больше какого-то из них?

Kazachenko1998 replied:
а разве НОД 5 и 10 не равен 5?
![image](https://cloud.githubusercontent.com/assets/22072948/18972690/d7ffea18-86a2-11e6-8a45-80b7cc81ff91.png)

gagarski replied:
Во-первых, 5 разве больше пяти?
Во-вторых, 5 и 10 не взаимно простые числа

Kazachenko1998 replied:
да, это было глупо... я забыл что мне нужно верное условие а не ложное...

ElenaLaufeyson committed:
Lesson3_task1
Исправила функции синуса и косинуса.

belyaev-mikhail commented:
recheck assignment

belyaev-mikhail commented:
recheck assignment

belyaev-mikhail commented:
recheck assignment

Kazachenko1998 committed:
lesson6.task1 проверка на ошибки

belyaev-mikhail commented:
recheck assignment

Kazachenko1998 commented:
Expected:
x=-632.0, y=-999.4227867839951, radius=0.5766311693979751  
x=-632.0, y=-999.9849946518009, radius=0.4792190693119849
S ==             0.562207868, и она пересекается
Actual:
x=-632.0, y=-999.9849946518009, radius=0.4792190693119849
x=-632.0, y=-1000.0                     , radius=0.3337186732548538
S ==             0.0150053482, и она пересекается
Разве действительный результат не верный?

Kazachenko1998 committed:
lesson6.task1 проверка очень сложной

ElenaLaufeyson committed:
Lesson4_task1
 Сделано еще несколько функций из 4 урока.

ElenaLaufeyson committed:
Lesson4_task1
Исправила функцию convert.

gagarski commented:
recheck all

gagarski commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -97,37 +126,74 @@ data class Line(val point: Point, val angle: Double) {
+    var point = PointInt(0, 1)
+    for (i in 0..circles.size - 1)
+        for (j in i + 1..circles.size - 1) {
+            if ((circles[i].distancee(circles[j]) <= min) && (circles[i].center != circles[j].center)) {
```
В данной задаче предполагается, что вы будете использовать функцию расстояния, которую вы определили в классе `Circle`. Ваше доопределённое для пересекающихся окружностей расстояние противоречит `distance`. Просто используйте `distance`.

gagarski commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -55,19 +56,30 @@ data class Circle(val center: Point, val radius: Double) {
+                (center.distance(other.center) - radius - other.radius)
+            else 0.0
+
+    fun distancee(other: Circle): Double =
```
Если уж определяете новые методы, то придумывайте им названия, отражающие их суть (а не лишь бы отличалось от другого метода). (Кстати, этот метод не нужен, см. комментарий к `findNearestCirclePair`).

gagarski commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -76,15 +88,32 @@ data class Segment(val begin: Point, val end: Point)
+
+fun diameter(vararg points: Point): Segment {
+    var max = -1.0
+    var maxPoint = PointInt(0, 0)
```
Никогда не храните в объектах класса то, для чего они не предназначены или не соответствует их названию. Класс `PointInt` представляет точку на плоскости с целочисленными координатами (судя по вашему же названию). Если вы хотите хранить произвольную пару чисел, используйте стандартный класс `Pair` или определите свой класс с соответствующим названием

gagarski commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -97,37 +126,74 @@ data class Line(val point: Point, val angle: Double) {
+    fun crossPoint(other: Line): Point {
+        var x = 0.0
+        var y = 0.0
+        if (Math.abs(Math.cos(angle)) <= 0.00000000001) {
```
Лучше записывать такие числа в экспоненциальной форме: `1e-11`

gagarski commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -97,37 +126,74 @@ data class Line(val point: Point, val angle: Double) {
+        var y = 0.0
+        if (Math.abs(Math.cos(angle)) <= 0.00000000001) {
+            x = point.x
+        } else if (Math.cos(other.angle) == 0.0) {
```
Почему выше сравнение с дельтой, а тут - строгое?

gagarski replied:
Можно ли обойтись одним особым случаем, когда прямые параллельны?

Kazachenko1998 replied:
если прямые параллельны то точки пересечения нет, а что делать в этом случае не сказано=> такого случая быть не должно!хотя конечно можно бросить исключение, но ведь это не просят... зачем делать лишнюю работу?

gagarski replied:
Справедливо, но, во-первых, замечание про дельту всё ещё актуально, а во-вторых, я писал скорее о том, что применив формулы тригонометрии можно обойтись без кода вроде `if (Math.abs(Math.cos(angle)) <= 0.00000000001)`

gagarski commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -97,37 +126,74 @@ data class Line(val point: Point, val angle: Double) {
+    if (circles.size < 2) throw IllegalArgumentException("IllegalArgumentException")
+    var min = circles[0].distance(circles[1])
+
+    var point = PointInt(0, 1)
```
`Pair`

Kazachenko1998 committed:
lesson6.task1 мелкие исправления

Kazachenko1998 committed:
lesson6.task1 доработан центр опис.окр

Kazachenko1998 committed:
lesson6.task1 очередные исправления в поиске окр.

Kazachenko1998 committed:
lesson6.task1 проверка на ошибки

Kazachenko1998 committed:
lesson6.task2 надеюсь работает

Kazachenko1998 committed:
lesson6.task2 исправленно

Kazachenko1998 commented:
а в чем ошибка?
ожидаемый результат - (пусто)
действительный результат- (пусто)
Ошибка- список пуст

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
> а в чем ошибка?
> ожидаемый результат - (пусто)
> действительный результат- (пусто)
> Ошибка- список пуст

Тесты для слона починил, `circleByThreePoints` пока не смотрел.

ElenaLaufeyson committed:
Lesson3_task1
Исправила синус и косинус

ElenaLaufeyson committed:
Lesson4_task1
Доделала 4 урок.

ElenaLaufeyson committed:
Lesson4_task1
Исправила ошибку в функции convertToString.

gagarski commented:
recheck assignment

gagarski commented:
recheck all

gagarski commented:
recheck all

belyaev-mikhail commented:
recheck all

belyaev-mikhail commented:
recheck all

belyaev-mikhail commented:
recheck all

belyaev-mikhail commented:
recheck all

gagarski commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -138,7 +197,28 @@ fun findNearestCirclePair(vararg circles: Circle): Pair<Circle, Circle> = TODO()
+    val x = -zx / (2 * z)
+    val y = zy / (2 * z)
+    val r = Math.sqrt(sqr(x - a.x) + sqr(y - a.y))
+    if ((x == Double.NaN) || (y == Double.NaN) || (r == Double.NaN) ||
```
Зачем всё это?

Во-первых, как ни странно, `Double.NaN != Double.NaN` (можете проверить). Нужно проверять на NaN методом `isNaN()`. Также есть метод `isInfinite()`.
Во-вторых, в задании не сказано, что делать при невозможности построить окружность, так что можно оставить всё как есть (вроде сейчас мы не должны подавать на вход такие данные).

ElenaLaufeyson committed:
Lesson4_task1
Исправила замечания к функциям convertToString и decimalFromString.

Kazachenko1998 committed:
lesson6.task1 исправленно

ElenaLaufeyson committed:
Lesson4_task1
Исправила замечания к функции convertToString.

Kazachenko1998 committed:
lesson7.task1 надеюсь работает...

ElenaLaufeyson committed:
Lesson5_task1
Сделала функции dateStrToDigit, dateDigitToStr, bestLongJump

Kazachenko1998 committed:
lesson7.task2 проверочка)))

ElenaLaufeyson committed:
Lesson5_task1
Сделала сложную функцию flattenPhoneNumber

gagarski commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -97,37 +120,71 @@ data class Line(val point: Point, val angle: Double) {
+        var x = 0.0
+        var y = 0.0
+        x = when {
+            (Math.abs(Math.cos(angle)) == Math.cos(Math.PI / 2)) -> point.x
```
Лучше `Math.tan(angle).isInfinite()`

Kazachenko1998 replied:
При замене (Math.abs(Math.cos(angle)) == Math.cos(Math.PI / 2)) на Math.tan(angle).isInfinite() программа ломается и не выводит ничего.

gagarski replied:
Что значит "не выводит ничего"? При каких входных данных?

gagarski commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +41,66 @@ interface Matrix<E> {
     }
 
-    override fun equals(other: Any?) = TODO()
+    override fun equals(other: Any?) =
```
Разве этих условий достаточно для сравнения матриц на равенство?

gagarski replied:
Скорее всего вы сгенерировали `equals` и `hashCode` автоматически до того, как добавили `list`. Попробуйте сгенерировать их заново или исправить вручную чтобы они выполняли корректное сравнение.

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented:
recheck all

gagarski commented on src/lesson7/task2/Matrices.kt:
```diff
@@ -59,7 +62,43 @@ operator fun Matrix<Int>.plus(other: Matrix<Int>): Matrix<Int> {
-fun generateSpiral(height: Int, width: Int): Matrix<Int> = TODO()
+fun turn(streamline: Int): Int {
+    return when (streamline) {
+        6 -> 2
```
Если вы хотите представить направления (вверх/вниз/влево/вправо) в программе, то используйте пречисления: https://kotlinlang.org/docs/reference/enum-classes.html

Kazachenko1998 committed:
lesson7.task1 переделал)))

Kazachenko1998 committed:
lesson7.task2 проверим...

Kazachenko1998 committed:
lesson7.task2 проверим...

Kazachenko1998 committed:
lesson7.task2 откуда взялись -1124145745 в findHoles???

Kazachenko1998 committed:
lesson7.task2

Kazachenko1998 commented:
может я конечно ошибаюсь, но программа должна работать верно в отличии от бота? ибо я не понимаю почему в последней посылке ожидается такой ответ, а не мой.

Kazachenko1998 committed:
lesson7.task2 c богом!

Kazachenko1998 committed:
lesson7.task2 исправления

gagarski commented:
recheck all

gagarski commented:
recheck all

Kazachenko1998 commented:
Почините, пожалуйста, бота в lesson7.task2/canOpenLock
И какие замечания нужно устранить по прошлым урокам?

gagarski commented:
> И какие замечания нужно устранить по прошлым урокам?

Можно посмотреть комментарии [тут](https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/68/files)

Kazachenko1998 committed:
Исправил замечания

Kazachenko1998 committed:
Первый вариант программы Centre

Kazachenko1998 committed:
lesson8.task1

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +183,43 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+    var number = n
+    while (i < number) {
+        i++
+        if (isPrime(i) == true)
```
Можно обойтись без проверки на простоту. Вы перебираете числа по возрастанию и делите число на все встречающиеся делители. После этого число просто не будет делиться на составные числа (поскольку они являются произведением простых меньше него). Например, вы нашли делители 2, 2, 3, 5 и поделили `number` на них. После этого число уже не будет делиться на 4, 6, 8, 12, 15, 16 и т. д.

`isPrime`, кстати уже был в третьем уроке в примерах.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -193,7 +268,12 @@ fun convertToString(n: Int, base: Int): String = TODO()
+fun decimal(digits: List<Int>, base: Int): Int {
+    var result = 0
+    for (i in (digits.size - 1) downTo 0)
+        result += (digits[i] * Math.pow(base.toDouble(), ((digits.size - 1) - i).toDouble())).toInt()
```
Не стоит переходить к числам с плавающей точкой в целочисленных задачах чтобы не столкнуться с ошибками округления. реализуйте `pow` для целых чисел, либо считайте множитель внутри цикла.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -214,7 +302,21 @@ fun decimalFromString(str: String, base: Int): Int = TODO()
+fun roman(n: Int): String {
+    var result = ""
+    var n2 = n
+    val rom = listOf("I", "IV", "V", "IX", "X", "XL", "L", "XC", "C", "CD", "D", "CM", "M")
```
Лучше список из пар, чем два списка, связанных общим индексом.

gagarski commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +237,33 @@ fun bestHighJump(jumps: String): Int = TODO()
+                if (parts[i - 1] == "-") result -= parts[i].toInt()
+                else result += parts[i].toInt()
+        }
+    } catch (e: IllegalArgumentException) {
```
Вы ловите `IllegalArgumentException` и тут же бросаете исключение такого же класса, но без сообщения. Нужен ли этот try/catch вообще?

gagarski commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +41,70 @@ interface Matrix<E> {
 
-    override fun equals(other: Any?) = TODO()
+    override fun toString(): String {
+        val sb = StringBuilder()
```
Ваша реализация `toString` записывает все числа из стоки матрицы подряд, без пробелов, запятых и пр. Хотя требований в задании к её реализации нет, лучше бы это исправить.

gagarski replied:
актуально

gagarski commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +41,70 @@ interface Matrix<E> {
+    }
+
+    override fun hashCode(): Int {
+        var result = 5
```
Хорошая функция `hashCode()` должна учитывать все значимые поля класса. 
Если вы её сгенерировали с помощью IDE автоматически, то, как минимум, сгенерируйте её заново, а лучше погуглите и почитайте, зачем он нужен и как его лучше реализовывать.

gagarski replied:
актуально

gagarski commented on src/lesson7/task2/Matrices.kt:
```diff
@@ -244,7 +445,78 @@ operator fun Matrix<Int>.times(other: Matrix<Int>): Matrix<Int> = TODO(this.toSt
+        for (j in 1..matrix.width)
+            if (matrix2[i, j] == 0) cell = Cell(i, j)
+    var i = -1
+    while (i < moves.size - 1) {
```
Для простого цикла по списку лучше `for`

gagarski commented on src/lesson7/task2/Matrices.kt:
```diff
@@ -244,7 +445,78 @@ operator fun Matrix<Int>.times(other: Matrix<Int>): Matrix<Int> = TODO(this.toSt
-fun fifteenGameMoves(matrix: Matrix<Int>, moves: List<Int>): Matrix<Int> = TODO()
+fun fifteenGameMoves(matrix: Matrix<Int>, moves: List<Int>): Matrix<Int> {
+    var faeled = true
+    for (n in 0..15) {
```
Перечитайте эту проверку на корректность ходов. Во-первых, это не работает (вы проверяете, что есть **хотя бы один** корректный ход), во-вторых, это кромешный ад. Нужно просто пройтись по матрице и проверить каждый из элементов на то, что они лежат в диапазоне 1..15. Если какой-либо элемент не лежит, можно сразу бросить исключение.

gagarski replied:
Снизу, кстати, есть нормальная проверка.

gagarski replied:
Неактуально

gagarski commented on src/lesson7/task2/Matrices.kt:
```diff
@@ -244,7 +445,78 @@ operator fun Matrix<Int>.times(other: Matrix<Int>): Matrix<Int> = TODO(this.toSt
+        if (faeled) throw IllegalStateException() else faeled = true
+    }
+    for (n in 0..moves.size - 1) if (moves[n] !in 1..15) throw IllegalStateException()
+    var matrix2 = createMatrix(matrix.height + 2, matrix.width + 2, 0)
```
Попробуйте обойтись без копирования матрицы в матрицу расширенного размера. Это не сильно упрощает логику.

В `matrix[0,0]` вы временно кладёте `Cell` перед обменом.

Ветки `when` лучше оформить как

``` kotlin
when {
    isInside(cell.row + 1, cell.column) && moves[i] == matrix2[Cell(cell.row + 1, cell.column)] -> {}
   // ...
}
```

`isInside` реализуйте отдельно.

gagarski replied:
Вроде же исправляли на занятии? А в репозитории нет.

gagarski commented:
recheck all

gagarski commented:
recheck all

Kazachenko1998 committed:
иправления

gagarski commented on src/lesson7/task2/Matrices.kt:
```diff
@@ -118,7 +221,23 @@ fun <E> rotate(matrix: Matrix<E>): Matrix<E> = TODO()
+fun isLatinSquare(matrix: Matrix<Int>): Boolean {
+    if (matrix.height != matrix.width) throw IllegalArgumentException()
+    var list = listOf<Int>()
+    var list2 = listOf<Int>()
```
`list2` лучше сделать локальной переменной внутри циклов, и инициализировать пустым списком при создании

mglukhikh commented:
recheck all

mglukhikh commented:
recheck all

mglukhikh commented:
Рад вам сообщить, что вы выполнили программу этого семестра с оценкой "ОТЛИЧНО" (можете подойти ко мне с зачёткой, и я вам эту оценку поставлю). При желании, можно порешать ещё какие-нибудь задачи.

mglukhikh commented:
Должен вас также поблагодарить за помощь в тестировании нашей инфраструктуры

Kazachenko1998 committed:
IC-2016.2.4 <Konstantin@Konstantin Create IntelliLang.xml, ide.general.xml, git.xml, Default.xml, project.default.xml, diff.xml, laf.xml, other.xml, jdk.table.xml, debugger.renderers.xml, applicationLibraries.xml, path.macros.xml, ui.lnf.xml, web-browsers.xml, keymap.xml, vcs.xml, editor.codeinsight.xml, gradle.run.settings.xml, colors.scheme.xml, notifications.xml, editor.xml, github_settings.xml, mavenVersion.xml, debugger.xml

Kazachenko1998 committed:
IC-2016.2.4 <Konstantin@Konstantin Update github_settings.xml

Kazachenko1998 committed:
IC-2016.2.4 <Konstantin@Konstantin Delete other.xml

Kazachenko1998 committed:
все доработки за последние 2-3 недели

Kazachenko1998 committed:
Merge remote-tracking branch 'origin/master'

Kazachenko1998 committed:
повтор...

Kazachenko1998 commented:
recheck all

Kazachenko1998 committed:
lesson7.task1

Kazachenko1998 committed:
IC-2016.2.4 <Konstantin@Konstantin Overwrite remote https://github.com/Kazachenko1998/KotlinAsFirst2016 to local

Kazachenko1998 committed:
IC-2016.2.4 <Konstantin@Konstantin Overwrite local to https://github.com/Kazachenko1998/KotlinAsFirst2016

Kazachenko1998 committed:
IC-2016.2.4 <Konstantin@Konstantin Overwrite local to https://github.com/Kazachenko1998/KotlinAsFirst2016

Kazachenko1998 committed:
IC-2016.2.4 <Konstantin@Konstantin Overwrite local to https://github.com/Kazachenko1998/KotlinAsFirst2016

Kazachenko1998 committed:
IC-2016.2.4 <Konstantin@Konstantin Update mavenVersion.xml

gagarski commented:
Проблема в том что вы (не без моего участия) добавили метод в интерфейс `Matrix`. При тестировании используется другая реализация этого интерфейса, в которой этого метода нет.

Вывод: изменения в интерфейсе могут привести к боли.

Можно добавить дефолтную реализацию метода `toList` в интерфейсе `Matrix` (она может использовать только методы этого интерфейса). Можно вообще убрать этот метод, поскольку вы его не используете.

Kazachenko1998 committed:
и снова попытка

Kazachenko1998 committed:
IC-2016.2.4 <Konstantin@Konstantin Create hg.xml

Kazachenko1998 committed:
lesson7.task1

Kazachenko1998 committed:
lesson7.task1

Kazachenko1998 committed:
lesson7.task1

Kazachenko1998 committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	src/lesson2/task1/IfElse.kt
#	src/lesson2/task2/Logical.kt
#	src/lesson3/task1/Loop.kt
#	src/lesson6/task1/Geometry.kt
#	src/lesson7/task1/Matrix.kt

Kazachenko1998 committed:
lesson7.task1

Kazachenko1998 commented:
Возникли небольшие трудности с IDEA(сильно сбились настройки) из-за чего столько посылок

Kazachenko1998 committed:
проверка

gagarski commented:
recheck all

gagarski commented:
recheck all

Kazachenko1998 committed:
IC-2016.2.4 <Konstantin@Konstantin Update mavenVersion.xml

gagarski commented:
recheck all

gagarski commented on src/lesson8/task1/Files.kt:
```diff
@@ -54,7 +53,22 @@ fun alignFile(inputName: String, lineLength: Int, outputName: String) {
+    var j = 0
+    substrings.toMutableList()
+    val result = mutableMapOf<String, Int>()
+    for (line in File(inputName).readLines())
```
`File().readText()`

gagarski commented:
По проваленным тестам в 8 уроке:

`countSubstrings` - исправил референсное решение, стал проходить.
`centerFile` - проверьте выполнение требования про число строк
`sibilants` - то же самое.

Kazachenko1998 committed:
IC-2016.3.2 <Konstantin@Konstantin Update path.macros.xml

Kazachenko1998 committed:
IC-2016.3.2 <Konstantin@Konstantin Update vcs.xml	Create parameter.hints.xml

Kazachenko1998 committed:
IC-2016.3.2 <Konstantin@Konstantin Update vcs.xml

Kazachenko1998 committed:
IC-2016.3.2 <Konstantin@Konstantin Update applicationLibraries.xml

Kazachenko1998 committed:
IC-2016.3.2 <Konstantin@Konstantin Update applicationLibraries.xml

Kazachenko1998 committed:
IC-2016.3.2 <Konstantin@Konstantin Update editor.codeinsight.xml

Kazachenko1998 committed:
IC-2016.3.2 <Konstantin@Konstantin Create toStringTemplates.xml

Kazachenko1998 committed:
IC-2016.3.2 <Konstantin@Konstantin Update diff.xml

Kazachenko1998 committed:
IC-2016.3.2 <Konstantin@Konstantin Update applicationLibraries.xml

Kazachenko1998 committed:
IC-2016.3.2 <Konstantin@Konstantin Update applicationLibraries.xml

Kazachenko1998 committed:
IC-2016.3.2 <Konstantin@Konstantin Update applicationLibraries.xml

Kazachenko1998 committed:
IC-2016.3.2 <Konstantin@Konstantin Update vcs.xml

Kazachenko1998 committed:
IC-2016.3.2 <Konstantin@Konstantin Overwrite remote https://github.com/Kazachenko1998/KotlinAsFirst2016 to local

Kazachenko1998 committed:
IC-2016.3.2 <Konstantin@Konstantin Update vcs.xml
